src/board.rs

```diff
@@ -0,0 +1,1021 @@
+        for (player, root) in r.iter().enumerate() {
+            if player != 0 {
+                // All tiles are already owned by the first player by default
+                for square in board.depth_first_search(*root).iter() {
+                    if let Ok(Square::Occupied(_, value)) = board.get(*square) {
+                        board.set(*square, player, value).expect(
+                            "A coordinate returned from a DFS should always be valid and settable",
+                        );
+                    }
+                }
+            }
+        }
+
+        Ok(board)
+    }
+
+    // TODO: generic board constructor that accepts a grid of squares with arbitrary values, as long as:
+    //  - the empty squares are fully connected
+    //  - there are at least 2 roots
+    //  - the roots are at empty squares
+
+    pub fn get(&self, position: Coordinate) -> Result<Square, &str> {
+        if position.y < 0 || position.x < 0 {
+            return Err("negative coordinates");
+        };
+        let x = position.x as usize;
+        let y = position.y as usize;
+
+        if y >= self.squares.len() {
+            Err("y-coordinate is too large for board height") // TODO: specify the coordinate and height
+        } else if x >= self.squares[0].len() {
+            Err("x-coordinate is too large for board width") // TODO: specify the coordinate and width
+        } else {
+            match self.squares[y][x] {
+                None => Err("Invalid position"),
+                Some(square) => Ok(square),
+            }
+        }
+    }
+
+    pub fn set(&mut self, position: Coordinate, player: usize, value: char) -> Result<(), &str> {
+        if position.y < 0 || position.x < 0 {
+            return Err("negative coordinates");
+        };
+        let x = position.x as usize;
+        let y = position.y as usize;
+
+        if player >= self.roots.len() {
+            Err("player does not exist") // TODO: specify the number of players and which player this is
+        } else if y >= self.squares.len() {
+            Err("y-coordinate is too large for board height") // TODO: specify the coordinate and height
+        } else if x >= self.squares[0].len() {
+            Err("x-coordinate is too large for board width") // TODO: specify the coordinate and width
+        } else {
+            match self.squares[y][x] {
+                Some(_) => {
+                    self.squares[y][x] = Some(Square::Occupied(player, value));
+                    Ok(())
+                }
+                None => Err("Can't set the value of a non-existant square"),
+            }
+        }
+    }
+
+    // TODO: safety on index access like get and set - ideally combine error checking for all 3
+    pub fn clear(&mut self, position: Coordinate) {
+        if self.squares[position.y as usize][position.x as usize].is_some() {
+            self.squares[position.y as usize][position.x as usize] = Some(Square::Empty);
+        }
+    }
```

Comment:
Generally for safety, reach for `vec.get(x)` over `vec[x]` — the following gives you full safety, and saves having to get that index a second time to set it:

```suggestion
pub fn clear(&mut self, position: Coordinate) {
        if let Some(pos) = self
            .squares
            .get_mut(position.y as usize)
            .and_then(|y| y.get_mut(position.x as usize))
        {
            *pos = Some(Square::Empty);
        }
    }
```